File: built-in plugins dsp/mb-synth-dubstep.cpp
/**
 * MB Dubstep Synth
 * Category : instrument
 * Type     : analog
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Aggressive dubstep sounds
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_SYNTH_DUBSTEP_H
#define MB_SYNTH_DUBSTEP_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbSynthDubstep : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-synth-dubstep";
    static constexpr const char* PLUGIN_NAME    = "MB Dubstep Synth";
    static constexpr const char* PLUGIN_TYPE    = "analog";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float drive = 0.7f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbSynthDubstep() = default;
    ~MbSynthDubstep() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.drive = std::clamp(params.drive, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Dubstep Synth
        return input;
    }
};

#endif // MB_SYNTH_DUBSTEP_H

File: built-in plugins dsp/mb-synth-dubstep.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbSynthDubstepPlugin: PluginDefinition = {
  id: "mb-synth-dubstep",
  slug: "mb-synth-dubstep",
  name: "MB Dubstep Synth",
  category: "instrument",
  type: "analog",
  version: "1.0.0",
  description: "Aggressive dubstep sounds",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "sawtooth", detune: 0, gain: 0.5 },
    { type: "square", detune: 0, gain: 0.5 },
  ],
  envelope: { attack: 0.001, decay: 0.1, sustain: 0.8, release: 0.1 },
  parameters: [
    {
      id: "drive",
      name: "Drive",
      type: "float",
      defaultValue: 0.7,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { drive: 0.7, volume: 0.8 },
};

export default MbSynthDubstepPlugin;
